### PR TITLE
Purchases: Reduxify Stored Cards on Edit Card Details page

### PR DESCRIPTION
**Pull request by stephanethomas**
This pull request is a follow-up of https://github.com/Automattic/wp-calypso/pull/6613, https://github.com/Automattic/wp-calypso/pull/6628, https://github.com/Automattic/wp-calypso/pull/6665, https://github.com/Automattic/wp-calypso/pull/6710 that updates the `Edit Card Details` page to use the global state tree to manage credit cards (stored during previous purchases or when a user added a new payment method to an existing purchase) instead of the former Flux store:

screenshot

It also fixes the prefilling of the card holder name that was broken.

#### Testing instructions
 
1. Run `git checkout update/reduxify-edit-card-details` and start your server, or open a [live branch](https://calypso.live/?branch=update/reduxify-edit-card-details)
2. Sandbox the store
3. Purchase e.g. a [plan](http://calypso.localhost:3000/plans) or a [site redirect](http://calypso.localhost:3000/domains/add/site-redirect) with a fake credit card
4. Open the [`Purchases` page](http://calypso.localhost:3000/purchases)
5. Click on the new purchase
6. Click on the `Edit Payment Method` action
7. Check that the `Name on Card` field contains the card holder name used in the checkout
8. Enter a new fake credit card (with a different name) and click the `Save Card` button
9. Check that you are redirected back to the `Manage Purchase` page with a success message
10. Check that you can see this new card on the `Billing History` page
11. Navigate back to the same `Edit Card Details` page
12. Check that the `Name on Card` field is prefilled in with the new name

You may want to check other payment methods, or type of purchases. You could also force a browser refresh on certain pages to make sure it doesn't make a difference.

#### Additional notes

I opted not to update the Redux selector to remap keys similarly to what the former assembler was doing because as you can see it here the only value we care is the card holder name. We use the keys provided by the API in any other places where we display credit cards. We can address that in a separate pull request though if most of us think it's worth it.

#### Reviews
 
- [x] Code
- [x] Product
- [x] Tests
 
@Automattic/sdev-feed

### CODE REVIEW

File: client/components/data/purchases/edit-card-details/index.jsx [hunk 1]
```diff
@@ -7,39 +7,38 @@ import React from 'react';
...
 	};
 }
 
 function isDataLoading( state ) {
 	return (
-		state.card.isFetching ||
 		! state.selectedPurchase.hasLoadedUserPurchasesFromServer ||
-		! state.hasLoadedSites
+		! state.hasLoadedSites ||
+		state.isFetchingStoredCards
```
Comment (stephanethomas) [inline client/components/data/purchases/edit-card-details/index.jsx]: I'm not super fan of referring here to a prop that is in fact set in the `EditCardDetails` component but managing it in this data component proved to be awkward (or maybe that's just me :) ).

File: client/components/data/purchases/edit-card-details/index.jsx [hunk 2]
```diff
@@ -55,24 +54,27 @@ const EditCardDetailsData = React.createClass( {
...
-				component={ this.props.component }
-				getStateFromStores={ getStateFromStores }
-				hasLoadedSites={ this.props.sites.fetched }
-				isDataLoading={ isDataLoading }
-				loadingPlaceholder={ this.props.loadingPlaceholder }
-				purchaseId={ this.props.purchaseId }
-				selectedSite={ this.props.sites.getSelectedSite() }
-				stores={ stores } />
+			<div>
+				<QueryStoredCards />
```
Comment (drewblaisdell) [inline client/components/data/purchases/edit-card-details/index.jsx]: Now that we're using `connect` to provide access to the global store through props, we don't need data components. Once this and #6647 are merged, we can remove this component entirely, so we should move any changes to it (like adding `QueryStoredCards`) into `EditCardDetails`.

**Comment by drewblaisdell** [comment]
This looks good. I had one note about moving the logic to display the loading placeholder and query the stored cards into `EditCardDetails` itself, in order to remove `EditCardDetailsData`.

**Comment by stephanethomas** [comment]
Thank you very much for the review @drewblaisdell. I updated this pull request per your suggestion. It is now ready for another review.

**Comment by fabianapsimoes** [comment]
I'm unable to open Manage Purchase in this branch. I see this error in the console when I click any items in the list in `/purchases`:

```TypeError: creditCard is undefined```

**Comment by stephanethomas** [comment]
Thanks for catching this error @fabianapsimoes. I've fixed the issue, this is ready for review again.

**Comment by fabianapsimoes** [comment]
Product :+1:

File: client/me/purchases/payment/edit-card-details/index.jsx [hunk 3]
```diff
@@ -71,27 +76,15 @@ const EditCardDetails = React.createClass( {
...
 		this.redirectIfDataIsInvalid();
 
 		recordPageView( 'edit_card_details', this.props );
 
 		let fields = formState.createNullFieldValues( this.fieldNames );
-		if ( this.props.card.data ) {
-			fields = this.mergeCard( this.props.card.data, fields );
+
+		if ( this.props.card ) {
+			fields.name = this.props.card.name;
```
Comment (drewblaisdell) [inline client/me/purchases/payment/edit-card-details/index.jsx]: Super minor, but `fields` should be a `const` now.

Comment (stephanethomas) [inline client/me/purchases/payment/edit-card-details/index.jsx]: Good catch!

**Comment by drewblaisdell** [comment]
I also tested this, and the code looks good to me. :+1:

**Pull request merged**